onnxruntime/core/optimizer/split_quickgelu_fusion.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/split_quickgelu_fusion.h"
+
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+using namespace onnxruntime;
+
+namespace {
+
+// pattern match on split, quickgelu and mult subgraph
+bool TrySplitQuickGeluMatch(Graph& graph, Node& start, Node*& split, Node*& quickgelu, Node*& mult) {
+  Node& node = start;
+  add = quickgelu = mult = nullptr;
+
+  // check node is split and has two outputs
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Split", {14}) ||
+      !graph_utils::IsSupportedProvider(node, {kCudaExecutionProvider, kRocmExecutionProvider}) ||
+      !optimizer_utils::CheckOutputEdges(graph, node, 2)) {
+    return false;
+  }
+
+  // check shape information is not available for input
+  Node& split_node = node;
+  NodeArg* input = split_node.MutableInputDefs()[0];
+  const TensorShapeProto* S = input->Shape();
+  if (S == nullptr || S->dim_size() < 1) {
+    return false;
+  }
+
+  // Split supports only float/float16/double/bfloat16 (OR MORE!!!!!????) - see ./onnxruntime/core/graph/contrib_ops/contrib_defs.cc
+  auto type_allowed = [](NodeArg* input) {
+    auto data_type = input->TypeAsProto()->tensor_type().elem_type();
+    if (data_type != ONNX_NAMESPACE::TensorProto_DataType_DOUBLE &&
+        data_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT16 &&
+        data_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+        data_type != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16) {
+      return false;
+    }
+    return true;
+  };
+  if (!type_allowed(input)) {
+    return false;
+  }
+
+  // Trying to find Split->QuickGelu->Mul Path
+  std::vector<const Node::EdgeEnd*> edges;
+  std::vector<graph_utils::EdgeEndToMatch> quickgelu_mul_path{
+      {0, 0, "QuickGelu", {1, 11, 13}, kOnnxDomain},
```

Comment:
the activation can be other elementwise operator instead of QuickGelu, maybe we can add a todo here to support other operators.